src/navigation/path_planner.py

```diff
@@ -0,0 +1,242 @@
+    target_vertex_idx: int
+
+    def __init__(self):
+        self.failure_zones: List[FailureZone] = []
+        self.visibility_graph: Graph = Graph(undirected=True)
+
+        self.source_pos: Point = None
+        self.target_pos: Point = None
+        self.path: List[Point] = None
+        self.target_vertex_idx: int = 0
+
+    def get_intermediate_target(self, source_pos: Point, target_pos: Point) -> Point:
+        """
+         Given a source position (the current position of the rover) and a target position
+         (the final destination), generates an intermediate target position after
+         calculating a clear path between source and destination.
+
+         This path is recalculated only when the target position changes, or a FailureZone
+         is added to the graph. Once the rover reaches this intermediate target,
+         complete_intermediate_target() should be called so that the path can update and
+         return the next target on the next call to get_intermediate_target().
+
+         Note: As the path is only recalculated when the target position changes, it is
+               assumed that the source_pos of the rover remains along this path. If you
+               you want to override this functionality and generate a path, call
+               generate_path() before calling this funciton.
+
+        :param source_pos:   Point object representing source position
+        :param target_pos:   Point object representing target position
+
+        :return:             A Point object representing the intermediate target
+        """
+        if (not self.path) or (not target_pos == self.target_pos):
+            self.generate_path(source_pos, target_pos)
+
+        if self.target_vertex_idx == len(self.path):
+            return self.path[-1]
+
+        return self.path[self.target_vertex_idx]
+
+    def complete_intermediate_target(self) -> None:
+        """
+        Marks the previously-returned intermediate target as complete.
+        """
+        assert self.path
+
+        if self.target_vertex_idx < len(self.path):
+            self.target_vertex_idx += 1
+
+    def is_path_complete(self) -> bool:
+        """
+        Returns whether the path has been fully traversed, which happens
+        if and only if the target_pos has been marked as completed.
+
+        :return: bool
+        """
+        return (self.path != None) and (self.target_vertex_idx == len(self.path))
+
+    def add_failure_zone(self, failure_zone: FailureZone) -> None:
+        """
+        Adds a failure zone to the visibility graph
+
+        :param failure_zone:    the failure zone to be added (as a FailureZone object)
+        """
+        # Step 1: add to list of failure zones
+        self.failure_zones.append(failure_zone)
+
+        # Step 2: remove all edges from current graph that cross this zone
+        graph = self.visibility_graph.get_data()
+        deleted_edges = []
+
+        for u, neighbors in graph.items():
```

Comment:
graph is a dict of vertex -> neighbor, where each neighbor object is a dict of vertex -> edge. Added a comment in the code as well. This format is controlled by the dijkstar library, so it can't really be controlled — seems to work pretty well tho